### PR TITLE
Seungheonoh/agora utils

### PR DESCRIPTION
**Pull request by SeungheonOh**
Check `CHANGELOG.md` for detail.

liqwid-labs/agora#147

### CODE REVIEW

**Comment by chfanghr** [comment]
IIRC https://github.com/Liqwid-Labs/liqwid-plutarch-extra/pull/39/commits/0ba9595f1a63d3c9eb56e9d2634a3a4ee17cb61c have been revert in one of the merge commits, so maybe your will have to make a revert-revert commit.

**Comment by SeungheonOh** [comment]
manually added the change logs because it was a merge that made the part go away, and I don't know how to fight with Git to revert that

**Comment by chfanghr** [comment]
Added a few more utils in https://github.com/Liqwid-Labs/agora/pull/149 and https://github.com/Liqwid-Labs/agora/pull/146 .

File: src/Plutarch/Extra/Maybe.hs
Comment (peter-mlabs) [inline src/Plutarch/Extra/Maybe.hs:202]: I think we could use better naming conventions with the other functions in this module (like `pfromDJust`.)

Comment (peter-mlabs) [reply src/Plutarch/Extra/Maybe.hs:202]: And likewise similar interfaces (passing in the error string vs not)

**Comment by SeungheonOh** [comment]
@chfanghr Added utility for liqwid-labs/agora#149. I don't think there's any addition for liqwid-labs/agora#146

File: src/Plutarch/Extra/MultiSig.hs
Comment (peter-mlabs) [inline src/Plutarch/Extra/MultiSig.hs:10]: Module name needs an update

File: src/Plutarch/Extra/IsData.hs
Comment (peter-mlabs) [inline src/Plutarch/Extra/IsData.hs:75]: Why do we need/want this? What is the point?

File: src/Plutarch/Extra/IsData.hs
Comment (peter-mlabs) [inline src/Plutarch/Extra/IsData.hs:78]: Likewise here -- what is the point of this?

File: src/Plutarch/Extra/Monoid.hs
Comment (peter-mlabs) [inline src/Plutarch/Extra/Monoid.hs:34]: Do we really need/want `punsafeCoerce` here?

Comment (peter-mlabs) [reply src/Plutarch/Extra/Monoid.hs:34]: Likewise for below

**Review by peter-mlabs — CHANGES_REQUESTED** [review]
Looks good except for the comments.

I'm also still not familiar with the SOP/deriving stuff, but if it seems to work that I don't have a problem merging.

File: src/Plutarch/Extra/IsData.hs
Comment (chfanghr) [reply src/Plutarch/Extra/IsData.hs:75]: For example: 

https://github.com/Liqwid-Labs/agora/blob/staging/agora/Agora/Proposal.hs#L319
https://github.com/Liqwid-Labs/agora/blob/staging/agora/Agora/Proposal.hs#L636

File: src/Plutarch/Extra/Maybe.hs
Comment (chfanghr) [reply src/Plutarch/Extra/Maybe.hs:202]: The intention here is to show a error message. I am not sure if this should be upstreamed to be honest.

File: src/Plutarch/Extra/Monoid.hs
Comment (chfanghr) [reply src/Plutarch/Extra/Monoid.hs:34]: It's fine to just use `punsafeCoerce` here, I suppose. Btw, using `pto` here will result in a type check error.

File: src/Plutarch/Extra/Maybe.hs
Comment (SeungheonOh) [reply src/Plutarch/Extra/Maybe.hs]: I renamed it to `passertPJust` and `passertPDJust`

File: src/Plutarch/Extra/IsData.hs
Comment (SeungheonOh) [reply src/Plutarch/Extra/IsData.hs:78]: Not sure. It's from Connor's PR.

**Comment by peter-mlabs** [comment]
Okay, I think this is good. I'll merge once CI passes

**Comment by SeungheonOh** [comment]
@peter-mlabs can this me merged?

File: src/Plutarch/Api/V1/ScriptContext.hs
Comment (kozross) [inline src/Plutarch/Api/V1/ScriptContext.hs:37]: Don't wildcard import constructors: spell out the ones you're using.

File: src/Plutarch/Api/V1/ScriptContext.hs
Comment (kozross) [inline src/Plutarch/Api/V1/ScriptContext.hs:316]: Missing `forall (s :: S) .` throughout.

File: src/Plutarch/Extra/Field.hs
Comment (kozross) [inline src/Plutarch/Extra/Field.hs:30]: Retrospective, but these all need kind signatures.

File: src/Plutarch/Extra/Field.hs
Comment (kozross) [inline src/Plutarch/Extra/Field.hs:46]: Same here.

File: src/Plutarch/Extra/Function.hs
Comment (kozross) [inline src/Plutarch/Extra/Function.hs:10]: Break this into multiple lines.

File: src/Plutarch/Extra/IsData.hs
Comment (kozross) [inline src/Plutarch/Extra/IsData.hs:28]: Again somewhat retroactive: no wildcard imports.

File: src/Plutarch/Extra/IsData.hs
Comment (kozross) [inline src/Plutarch/Extra/IsData.hs:187]: Somewhat retrospective, but we need a kind signature on `s` here.

File: src/Plutarch/Extra/List.hs
Comment (kozross) [inline src/Plutarch/Extra/List.hs:30]: I thought we were removing uses of this?

File: src/Plutarch/Extra/Map.hs
Comment (kozross) [inline src/Plutarch/Extra/Map.hs:22]: Same again.

File: src/Plutarch/Extra/Maybe.hs
Comment (kozross) [inline src/Plutarch/Extra/Maybe.hs:216]: Kind signatures missing.

File: src/Plutarch/Extra/Monoid.hs
Comment (kozross) [reply src/Plutarch/Extra/Monoid.hs:34]: I don't think this is too bad here, but I feel it'd probably be better to `pmatch` then rewrap. Newtypes are invisible at runtime.

File: src/Plutarch/Extra/MultiSig.hs
Comment (kozross) [inline src/Plutarch/Extra/MultiSig.hs:25]: No wildcard imports.

File: src/Plutarch/Extra/MultiSig.hs
Comment (kozross) [inline src/Plutarch/Extra/MultiSig.hs:160]: Kind signature for `s` missing.

File: src/Plutarch/Extra/Other.hs
Comment (kozross) [inline src/Plutarch/Extra/Other.hs:33]: Slightly retrospective, but no wildcard imports please.

**Review by kozross — CHANGES_REQUESTED** [review]

**Comment by SeungheonOh** [comment]
This is not dependent on #43. We need to hold merge until we are done with it.

**Comment by peter-mlabs** [comment]
Should be good to go, but @chfanghr wants to check `DerivePDataFieldsViaDataList` a bit more before merging to `main`. Give the thumbs up when we're good!

**Review by peter-mlabs — APPROVED** [review]
Got the thumbs up, so this is getting merged :partying_face:

**Pull request merged**
